core/arch/arm/include/arm.h

```diff
@@ -107,7 +107,11 @@
 static inline __noprof uint64_t barrier_read_counter_timer(void)
 {
 	isb();
+#ifdef CFG_CORE_SEL2_SPMC
+	return read_cntvct();
+#else
 	return read_cntpct();
+#endif
```

Comment:
Shouldn't we always use CNTVCT instead? If I understand correctly the virtual timer is the same as the physical one when EL2 is not enabled anyways.

Reply:
Good point, I'll test a bit. If stuff works as expected I'll create a separate PR for that and drop a few patches in this PR.

Reply:
It looks like it could work on ARMv8, but on ARMv7 we need to initialize CNTVOFF if the Virtualization Extensions are implemented.
For now it seems easier to continue to use CNTPCT when we don't have a secure hypervisor.

Reply:
> It looks like it could work on ARMv8, but on ARMv7 we need to initialize CNTVOFF if the Virtualization Extensions are implemented.

This is already taken care of by `CFG_INIT_CNTVOFF`.
https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/sm/sm_a32.S#L338-L367

Reply:
Thanks for the link, I had forgot about that.

I've investigated further and it seems that we're actually covered for ARMv7. Here's what I've found out.

ARMv7:
- CNTVOFF is always 0 in secure world from CNTVCT point of view

ARMv8:
- If EL2 isn't implemented CNTVOFF is always 0 in from CNTVCT point of view
- CNTVOFF_EL2 can be saved and restored by el2_sysregs_context_save() and el2_sysregs_context_restore(), but isn't in OPTEED

For ftrace we need to set CNTKCTL_PL0VCTEN instead of CNTKCTL_PL0PCTEN where needed.

For ARMv8 we need to change the the OP-TEE Dispatcher, OPTEED, in TF-A at EL3. TF-A should initialize CNTVOFF_EL2 to 0 and then save and restore it. I'm a bit reluctant to do this as we risk breaking downstream configurations.

Did I miss something?

Reply:
I tried to wrap my head around all this and I must admit I am not familiar enough with them to decide what is the best way. My impression is that PCT is normally used by EL3, EL2 and/or S-EL2 while EL1 and S-EL1 (and EL0 and S-EL0) should rather use VCT the idea being they are the same without virtualization, while VCT is needed to support some virtualization use cases (VM migration etc.). That is why I proposed to use VCT unless there are good reasons to use PCT.

> For ftrace we need to set CNTKCTL_PL0VCTEN instead of CNTKCTL_PL0PCTEN where needed.

OK

> For ARMv8 we need to change the the OP-TEE Dispatcher, OPTEED, in TF-A at EL3. TF-A should initialize CNTVOFF_EL2 to 0 and then save and restore it. I'm a bit reluctant to do this as we risk breaking downstream configurations.

Saving and restoring should not be an issue, right? As for initialization, why would an earlier step than TFA initialize VOFF to anything other than zero? That would be a weird thing to do which could be addressed by conditional compilation in TF-A I suppose.

As I said I'd be happy to hear from someone with a good knowledge of virtualization and of these timers... @lorc maybe?

Reply:
Well, S-EL2 is a quite new thing. But in case of regular EL2 - @jforissier is right. Linux for instance always uses virtual timer, regardless if virtualization is enabled. 
I believe, that for S-EL2 the same approach should be used.

Reply:
For now I think it's better to use PCT when OP-TEE is physical and VCT when there's a S-EL2. If we want to change this to always use VCT later on, but that will be an ABI change between S-EL1 and EL3.
This PR is about the case with a S-EL2 in the picture and is not supposed to change the behaviour for the case without S-EL2.

Reply:
> This PR is about the case with a S-EL2 in the picture and is not supposed to change the behaviour for the case without S-EL2.

That's a good argument.